### PR TITLE
Enable is_angular and weights in loss.mse

### PR DESCRIPTION
**Pull request by edoyango**
Hi @tennlee 

I didn't see you push any new stuff, so I started working on `loss.mse`. This PR gets `loss.mse` working with `is_angular` and `weights`. Tests have been added to check that results match for varying `np.ndarray` and `torch.Tensor` inputs (with multiple device tests for the latter - conditional on whether torch is available as well as the relevant device). `pd.Series` inputs doesn't support `is_angular` and `weights` since it doesn't follow Python Array API. Hence, those tests are skipped for now.

Just wanted to put this up to see if this matches what you had in mind.

## Final Checks:

 - [X] ~~If no generative AI was used, then tick this box~~ Claude Sonnet 5 used to organise tests (reduce duplication etc.) and get the code to pass pre-commit checks.

### CODE REVIEW

**Comment by edoyango** [comment]
@tennlee I implemented some of your suggestions for the tests RE: reducing duplication and ensuring data is consistent between `test_standard.py` and `test_loss_continuous.py`. The data is initialized as xr.DataArray in `tests/continuous/mse_bias_test_data.py` and imported into both meaning changes in `tests/continuous/mse_bias_test_data.py` get propagated into both test suites. 

I also implemented additive_bias. When adding the additive_bias tests, I learnt that `torch.mean` propagates NaNs to the result, whereas xarray mean skips the values. Hence, changes were made to replicate this behaviour for non xarray objects (with comments flagging this).

Let me know if this is closer to your vision.

File: tests/categorical/test_contingency.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 import xarray as xr
 
 import scores
+import scores.categorical
```
Comment (tennlee) [inline tests/categorical/test_contingency.py]: You don't need to import all the subpackages, the top-level package imports them all by default.

File: tests/loss/test_loss_continuous.py [hunk 2]
```diff
@@ -5,91 +5,165 @@
...
+
+    _SKIP_TORCH_TESTS = False
+except ModuleNotFoundError:
+    _SKIP_TORCH_TESTS = True
 
 import scores
+import scores.continuous
+import scores.loss
+from tests.continuous.mse_bias_test_data import (
+    BIAS_FCST_DA,
```
Comment (tennlee) [inline tests/loss/test_loss_continuous.py:23]: Yes, something like this is great. I

File: src/scores/processing/aggregation.py [hunk 3]
```diff
@@ -144,11 +150,22 @@ def _weighted_mean(values, weights, reduce_dims=None):
...
     # Else attempt to handle with array compatibility
+    # xarray mean behaviour to be replicated:
+    # * when encountering nan in `values`, element is excluded from both numerator
+    #   and corresponding element in `weights` is excluded from denominator.
+    # * when encountering nan in `weights`, function raises exception.
     else:
         xp = array_api_compat.array_namespace(values, weights)
-        weighted_error = xp.mul(values, weights)
+        # TODO: check that weights are >= 0
+        if bool(xp.any(xp.isnan(weights))):
```
Comment (tennlee) [inline src/scores/processing/aggregation.py]: Can we use scores.utils.check_weights, or perhaps implement scores.loss.utils.check_weights? I think the check is for the same thing - non-negative, at least one positive (i.e. nonzero) value.

File: src/scores/loss/_loss_continuous.py
Comment (tennlee) [inline src/scores/loss/_loss_continuous.py:5]: I'm not sure how to add type hints for array_api_compat, we should look into it.

File: src/scores/loss/_loss_continuous.py [hunk 4]
```diff
@@ -7,6 +6,7 @@ def mse(
     obs,
     *,  # Force keywords arguments to be keyword-only
     is_angular: bool = False,
+    weights=None,
```
Comment (tennlee) [inline src/scores/loss/_loss_continuous.py:9]: Something I wasn't sure of for early implementations was whether to do what's here now, or also allow reduce and preserve dim arguments, but raise a NotImplementedError if a value is passed in, together with an explanation message about why, and that the feature is intended for a future release. I think it's best to just not have them until we actually can support them, but I'd be interested in feedback on this point.

**Comment by tennlee** [comment]
Hi,

This looks very much on the right track, thanks for the great work. I've added atomic comments in various places. At a higher level, here are some additional reflections.

1. We will want a notebook demonstrating the use of the loss functions in training, even if it's not on a full dataset. I can make some suggestions here. I would probably suggest training an autoencoder on some open data, just doing a few samples should be sufficient to demonstrate the point. Let me know if you'd rather I did this, or whether you'd like to do it.
2. `scores` is a bit different to other packages in that it brings in the entire API at once and doesn't need subpackage imports. While it's still quick enough, I think we'll just stick with it, but it means you don't need some of the import statements you have (though they are harmless)
3. We will need a plan for rolling this into `develop`. We probably need some manual testing first among a few people. We should also get review from one of the scientists to make sure they are equally as happy with the implementation. They might suggest additional tests they think are important, I'm not sure.
4. I'm not sure whether to go with implementing just these two methods for now (probably not, but open to argument) or whether to add say tw_huber_loss as a more complex loss function that will probably have interested users, and prove the pattern with that more complex example.
5. CRPS is another valuable loss function. 
6. Adding reduce_dims and preserve_dims is possible, but in my mind it can wait for a subsequent phase of the work
7. It might be interesting to find some papers from the weather field which have used these loss functions and list them in the documentation for interested readers who might be wondering which loss functions have been used to good effect in the literature


These comments go a bit beyond feedback on your work, and extend into a discussion on what to bite off in the first round.

A practical way forward might be to tidy up the work so far, and merge it into a feature branch on the main repository that we can use for a more structured PR process with finer-grained commits from this point.

We can talk more shortly. Thanks very much for the work. Shaping up well.

**Comment by nicholasloveday** [comment]
Suggest starting with the consistent scoring functions rather than the threshold weighted scores. https://github.com/nci/scores/blob/develop/src/scores/continuous/consistent_impl.py

The threshold weighted scores call the consistent functions under the hood.

**Comment by nicholasloveday** [comment]
I suggest allowing a `preserve_dims="all"` arg. This will mean that with the two options: 1 reduce all, and. 2 preserve all, a user can then replicate all preserve/reduce behaviour in scores if they add an additional line of code.

**Comment by nicholasloveday** [comment]
> CRPS is another valuable loss function.

We should also consider implementing the "almost fair" CRPS

File: tests/categorical/test_contingency.py (hunk 1 above)
Comment (edoyango) [reply tests/categorical/test_contingency.py]: Yeah i didn't really like it, but the `ty-check` pre-commit failed if I didn't have an explicit import

```
warning[possibly-missing-submodule]: Submodule `continuous` might not have been imported
  --> tests/loss/test_loss_continuous.py:68:13
   |
68 |             scores.continuous.mse(
   |             ^^^^^^^^^^^^^^^^^
   |
help: Consider explicitly importing `scores.continuous`

warning[possibly-missing-submodule]: Submodule `loss` might not have been imported
  --> tests/loss/test_loss_continuous.py:97:21
   |
97 |     device_result = scores.loss.mse(
   |                     ^^^^^^^^^^^
   |
help: Consider explicitly importing `scores.loss`
```

I guess an alternative would be to import the stuff i need explicitly instead of pulling them in from `scores`. e.g. `from scores.loss import mse`. Would your rather that instead? or ignore the ty-check pre-commit for now?

File: src/scores/loss/_loss_continuous.py
Comment (edoyango) [reply src/scores/loss/_loss_continuous.py:5]: i had a quick look and supporting doesn't seem great... found this package by the same org as `array_api_compat` but it says it's still in dev https://github.com/data-apis/array-api-typing.

File: src/scores/processing/aggregation.py (hunk 3 above)
Comment (edoyango) [reply src/scores/processing/aggregation.py]: Can do, although I reckon we could implement it into the same function by adding an `elif` branch which does the check for array compat objects. What do you think?

File: src/scores/loss/_loss_continuous.py (hunk 4 above)
Comment (edoyango) [reply src/scores/loss/_loss_continuous.py:9]: I agree with the `NotImplementedError` approach. More informative than an unexpected keyword argument error.

File: src/scores/processing/aggregation.py (hunk 3 above)
Comment (edoyango) [reply src/scores/processing/aggregation.py]: Hm i had another look and I'm not sure why I added this TODO. The existing DataArray/Dataset branch doesn't seem to care if there are no positive numbers in the weights. Will remove the TODO.

File: tests/categorical/test_contingency.py (hunk 1 above)
Comment (edoyango) [reply tests/categorical/test_contingency.py]: got rid of the subpackage imports because i don't like them either. will ignore ty-check for now.

**Comment by edoyango** [comment]
@tennlee 

I've had a go at adding preliminary python array api support for `tw_huber_loss`. I've only ported over the single test that doesn't use `preserve_dims` from `tests.continuous.test_threshold_weighted`. Once `preserve_dims="all"` is implemented, I'll port the remaining tests over.

As with the `mse` tests, I've moved the overlapping test data to a separate module to be reused by both `tests.loss` and `tests.continuous`.

Some feedback I would like: I've added the `array_ops` module which include `where, nanmin, nanmax` ops. The motivation for this:
* `where` is used frequently, and having the dispatch in a central module is imo neater than each function doing the dispatch within.
* `DataArray/Dataset.min/max` behave like `np.nanmin/max` (skipping over nans), but `torch` doesn't implement this. Again, having the central module significantly reduces duplication in `threshold_weighted_impl` since this behaviour is relied on a few of the aux functions.

Other ops that affect `mse` could be added (like `nanmean`), but I thought I'd get another opinion before fully committing.

**Comment by edoyango** [comment]
Also lmk if it would be easier to separate this out into a different PR, since there's probably a lot of changes now
